File: sdk/csharp/src/AnyVali/SensitiveData.cs
namespace AnyVali;

/// <summary>Explicit encrypted-storage helpers for sensitive schema nodes.</summary>
public static class SensitiveData
{
    public static ParseResult SafeParseEncrypted(Schema schema, object? data)
    {
        var ctx = new ValidationContext { SensitiveMode = "encrypted" };
        var output = schema.RunPipeline(data, ctx);
        return ctx.Issues.Count > 0 ? ParseResult.Fail(ctx.Issues) : ParseResult.Ok(output);
    }

    public static object? Encrypt(
        Schema schema,
        object? data,
        Func<IReadOnlyList<object>, object?, object?> transform)
    {
        var encrypted = Transform(schema, schema.Parse(data), "encrypt", transform);
        var checkedData = SafeParseEncrypted(schema, encrypted);
        if (!checkedData.Success) throw new ValidationError(checkedData.Issues);
        return checkedData.Data;
    }

    public static object? Decrypt(
        Schema schema,
        object? data,
        Func<IReadOnlyList<object>, object?, object?> transform)
    {
        var checkedData = SafeParseEncrypted(schema, data);
        if (!checkedData.Success) throw new ValidationError(checkedData.Issues);
        return schema.Parse(Transform(schema, checkedData.Data, "decrypt", transform));
    }

    private static object? Transform(
        Schema schema,
        object? data,
        string mode,
        Func<IReadOnlyList<object>, object?, object?> transform)
    {
        var ctx = new ValidationContext
        {
            SensitiveMode = mode,
            SensitiveTransform = transform,
            SensitiveCache = new Dictionary<string, object?>(),
        };
        var output = schema.RunPipeline(data, ctx);
        if (ctx.Issues.Count > 0) throw new ValidationError(ctx.Issues);
        return output;
    }
}

File: sdk/php/src/SensitiveData.php
<?php

declare(strict_types=1);

namespace AnyVali;

final class SensitiveData
{
    private function __construct()
    {
    }

    public static function safeParseEncrypted(Schema $schema, mixed $data): ParseResult
    {
        return $schema->safeParse($data, new ValidationContext(sensitiveMode: 'encrypted'));
    }

    public static function encrypt(Schema $schema, mixed $data, callable $transform): mixed
    {
        $encrypted = self::transform($schema, $schema->parse($data), 'encrypt', $transform);
        $result = self::safeParseEncrypted($schema, $encrypted);
        if (!$result->success) throw new ValidationError($result->issues);
        return $result->value;
    }

    public static function decrypt(Schema $schema, mixed $data, callable $transform): mixed
    {
        $encrypted = self::safeParseEncrypted($schema, $data);
        if (!$encrypted->success) throw new ValidationError($encrypted->issues);
        return $schema->parse(self::transform($schema, $encrypted->value, 'decrypt', $transform));
    }

    private static function transform(Schema $schema, mixed $data, string $mode, callable $transform): mixed
    {
        $result = $schema->safeParse($data, new ValidationContext(
            sensitiveMode: $mode,
            sensitiveTransform: $transform,
            sensitiveCache: new \ArrayObject(),
        ));
        if (!$result->success) throw new ValidationError($result->issues);
        return $result->value;
    }
}
